### PR TITLE
feat: add in hybrid slash commands

### PR DESCRIPTION
**Pull request by AstreaTSS**
## Pull Request Type

- [x] Feature addition
- [ ] Bugfix
- [ ] Documentation update
- [ ] Code refactor
- [ ] Tests improvement
- [ ] CI/CD pipeline enhancement
- [ ] Other: [Replace with a description]

## Description
Adds (back) hybrid slash commands, which makes it easy to make and use a prefixed command from a standard slash command. This is based off the old NAFF implementation, though improvements have been made where possible.

## Changes
It's hybrid commands.
- Add in a new extension to deal with hybrid commands.
  - Add in a manager for ease-of-injection.
  - Add in `HybridSlashCommand` and the command that transforms a slash command to a prefixed command.
  - Add in `HybridContext` for nice interop between the two types of commands.

## Test Scenarios
```python
import interactions
from interactions.ext import prefixed_commands as prefixed
from interactions.ext import hybrid_commands as hybrid

prefixed.setup(bot, generate_prefixes=generate_prefixes)
hybrid.setup(bot)

@hybrid.hybrid_slash_command()
@interactions.slash_option(
    "msg",
    "The message to send",
    opt_type=interactions.OptionType.INTEGER,
    required=False,
)
async def test(ctx: hybrid.HybridContext, msg: int | None = 1):
    await ctx.reply(str(msg * 5))
```

## Python Compatibility

- [ ] I've ensured my code works on Python `3.10.x`
- [x] I've ensured my code works on Python `3.11.x`

## Checklist

- [x] I've run the `pre-commit` code linter over all edited files
- [x] I've tested my changes on supported Python versions
- [ ] I've added tests for my code, if applicable
- [x] I've updated / added  documentation, where applicable

### CODE REVIEW

**Comment by AstreaTSS** [comment]
Ready for testing. Do note the extent of these changes - as with the original hybrid command PR from a while back, testing everything is impossible and there's likely plenty of bugs right now.

**Review by AstreaTSS — COMMENTED** [review]
Bumping. Two developers from the Discord server have said it's stable for them - I think this is safe to merge.

**Comment by AstreaTSS** [comment]
Rebased! Also did a tiny adjustment to the title of the hybrid command section to note how prefixed commands are often called text commands.

File: interactions/ext/hybrid_commands/context.py [hunk 1]
```diff
@@ -0,0 +1,386 @@
...
+    def from_prefixed_context(cls, ctx: prefixed.PrefixedContext) -> Self:
+        # this is a "best guess" on what the permissions are
+        # this may or may not be totally accurate
+        if hasattr(ctx.channel, "permissions_for"):
+            app_permissions = ctx.channel.permissions_for(ctx.guild.me)  # type: ignore
+        elif ctx.channel.type in {10, 11, 12}:  # it's a thread
+            app_permissions = ctx.channel.parent_channel.permissions_for(ctx.guild.me)  # type: ignore
+        else:
+            # this is what happens with interaction contexts in dms
+            app_permissions = Permissions(0b1111100110001000000)
```
Comment (i0bs) [inline interactions/ext/hybrid_commands/context.py]: what in god's fucking name

Comment (i0bs) [reply interactions/ext/hybrid_commands/context.py]: Can we extract the bytes to a flag, or comment on what the permissions are exactly?

File: interactions/ext/hybrid_commands/hybrid_slash.py [hunk 2]
```diff
@@ -0,0 +1,593 @@
...
+        self.type_to_convert = type_to_convert
+
+    async def convert(self, ctx: BaseContext, arg: str) -> Any:
+        return self.type_to_convert(arg)
+
+
+class BoolConverter(Converter):
+    async def convert(self, ctx: BaseContext, argument: str) -> bool:
+        lowered = argument.lower()
+        if lowered in {"yes", "y", "true", "t", "1", "enable", "on"}:
```
Comment (i0bs) [inline interactions/ext/hybrid_commands/hybrid_slash.py:75]: Maybe we should take these literal sets and move them to variable declarations so we or developers can modify them

File: interactions/ext/hybrid_commands/hybrid_slash.py [hunk 3]
```diff
@@ -0,0 +1,593 @@
...
+
+
+class StringLengthConverter(Converter[str]):
+    def __init__(self, min_length: Optional[int], max_length: Optional[int]) -> None:
+        self.min_length = min_length
+        self.max_length = max_length
+
+    async def convert(self, ctx: BaseContext, argument: str) -> str:
+        if self.min_length and len(argument) < self.min_length:
+            raise BadArgument(f'The string "{argument}" is shorter than' f" {self.min_length} character(s).")
```
Comment (i0bs) [inline interactions/ext/hybrid_commands/hybrid_slash.py]: Why not keep it to one f-string?

```suggestion
            raise BadArgument(f'The string "{argument}" is shorter than {self.min_length} character(s).')
```

File: interactions/ext/hybrid_commands/hybrid_slash.py [hunk 4]
```diff
@@ -0,0 +1,593 @@
...
+        elif self.max_length and len(argument) > self.max_length:  # noqa: RET506
+            raise BadArgument(f'The string "{argument}" is longer than' f" {self.max_length} character(s).")
```
Comment (i0bs) [inline interactions/ext/hybrid_commands/hybrid_slash.py]:
```suggestion
            raise BadArgument(f'The string "{argument}" is longer than {self.max_length} character(s).')
```

**Review by i0bs — CHANGES_REQUESTED** [review]

File: interactions/ext/hybrid_commands/hybrid_slash.py (hunk 3 above)
Comment (AstreaTSS) [reply interactions/ext/hybrid_commands/hybrid_slash.py]: weird black moment

File: interactions/ext/hybrid_commands/context.py (hunk 1 above)
Comment (AstreaTSS) [reply interactions/ext/hybrid_commands/context.py]: I suppose, though I think I may just make it `None` for parity with slash commands.

File: interactions/ext/hybrid_commands/hybrid_slash.py (hunk 2 above)
Comment (AstreaTSS) [reply interactions/ext/hybrid_commands/hybrid_slash.py:75]: Like the idea, but as of right now, that converter's meant to have parity with prefixed commands. I'll add it in another PR for prefixed commands though - I do like it.

**Review by i0bs — APPROVED** [review]

**Pull request merged**